src/daft-core/src/array/ops/list.rs

```diff
@@ -244,6 +252,128 @@ fn list_sort_helper_fixed_size(
 }
 
 impl ListArray {
+    pub fn value_counts(&self) -> DaftResult<MapArray> {
+        struct IndexRef {
+            index: usize,
+            hash: u64,
+        }
+
+        impl std::hash::Hash for IndexRef {
+            fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+                self.hash.hash(state);
+            }
+        }
+
+        let original_name = self.name();
+
+        let hashes = self.flat_child.hash(None)?;
+
+        let flat_child = self.flat_child.to_arrow();
+        let flat_child = &*flat_child;
+
+        let is_equal = build_is_equal(
+            flat_child, flat_child, true, // todo: should nans be considered equal?
+            true,
+        )?;
+
+        let is_valid = build_is_valid(flat_child);
+
+        let key_type = self.flat_child.data_type().clone();
+        let count_type = DataType::UInt64;
+
+        let mut include_mask = Vec::with_capacity(self.flat_child.len());
+        let mut count_array = Vec::new();
+
+        let mut offsets = Vec::with_capacity(self.len());
+
+        offsets.push(0_i64);
+
+        let mut map: IndexMap<IndexRef, u64, IdentityBuildHasher> = IndexMap::default();
+        for range in self.offsets().ranges() {
+            map.clear();
+
+            for index in range {
+                let index = index as usize;
+                if !is_valid(index) {
+                    include_mask.push(false);
+                    // skip nulls
+                    continue;
+                }
+
+                let hash = hashes.get(index).unwrap();
+
+                let entry = map
+                    .raw_entry_mut_v1()
+                    .from_hash(hash, |other| is_equal(other.index, index));
+
+                match entry {
+                    RawEntryMut::Occupied(mut entry) => {
+                        include_mask.push(false);
+                        *entry.get_mut() += 1;
+                    }
+                    RawEntryMut::Vacant(vacant) => {
+                        include_mask.push(true);
+                        vacant.insert(IndexRef { hash, index }, 1);
+                    }
+                }
+            }
+
+            // indexmap so ordered
```

Comment:
Cleanup comment or explain why we want the result to be ordered and that indexmap is ordered.

Reply:
cae76fd0743d48816ffbb03caa2fe1d6e30630c6